engine/schema/src/main/java/com/cloud/upgrade/SystemVmTemplateRegistration.java

```diff
@@ -0,0 +1,769 @@
+            final String templateName = UUID.randomUUID().toString();
+            Date created = new Date(DateUtil.currentGMTTime().getTime());
+            SystemVMTemplateDetails details = new SystemVMTemplateDetails(templateName, NewTemplateNameList.get(hypervisor), created,
+                    newTemplateUrl.get(hypervisor), newTemplateChecksum.get(hypervisor), hypervisorImageFormat.get(hypervisor), hypervisorGuestOsMap.get(hypervisor), hypervisor, storeUrlAndId.second());
+            templateId = createTemplateObjectInDB(conn, details);
+            if (templateId == null) {
+                throw new CloudRuntimeException(String.format("Failed to register template for hypervisor: %s", hypervisor.name()));
+            }
+            details.setId(templateId);
+            String destTempFolderName = String.valueOf(templateId);
+            String destTempFolder = PARENT_TEMPLATE_FOLDER + PARTIAL_TEMPLATE_FOLDER + destTempFolderName;
+            details.setInstallPath(PARTIAL_TEMPLATE_FOLDER + destTempFolderName + "/" + templateName + "." + hypervisorImageFormat.get(hypervisor).fileExtension);
+            createTemplateStoreRefEntry(conn, details);
+            String storageScriptsDir = "scripts/storage/secondary";
+            String setupTmpltScript = Script.findScript(storageScriptsDir, "setup-sysvm-tmplt");
+            if (setupTmpltScript == null) {
+                throw new ConfigurationException("Unable to find the createtmplt.sh");
+            }
+            Script scr = new Script(setupTmpltScript, SCRIPT_TIMEOUT, LOGGER);
+            scr.add("-u", templateName);
+            scr.add("-f", templatesPath + fileNames.get(hypervisorAndTemplateName.first()));
+            scr.add("-h", hypervisorAndTemplateName.first().name().toLowerCase(Locale.ROOT));
+            scr.add("-d", destTempFolder);
+            String result = scr.execute();
+            if (result != null) {
+                String errMsg = String.format("failed to create template: %s ", result);
+                LOGGER.error(errMsg);
+                throw new CloudRuntimeException(errMsg);
+            }
+            details.setInstallPath(PARTIAL_TEMPLATE_FOLDER + destTempFolderName + "/" + templateName + "." + hypervisorImageFormat.get(hypervisor).fileExtension);
+            readTemplateProperties(destTempFolder + "/template.properties", details);
+            details.setUpdated(new Date(DateUtil.currentGMTTime().getTime()));
+            updateDb(conn, details);
+            Map<String, String> configParams = new HashMap<>();
+            configParams.put(SystemVmTemplateRegistration.routerTemplateConfigurationNames.get(hypervisorAndTemplateName.first()), hypervisorAndTemplateName.second());
+            configParams.put("minreq.sysvmtemplate.version", CS_MAJOR_VERSION + "." + CS_TINY_VERSION);
+            updateConfigurationParams(conn, configParams);
+            updateSystemVMEntries(conn, templateId, hypervisorAndTemplateName);
```

Comment:
There are several process here, we could extract some to minor methods. Also, we could add unit tests to these methods.